src/lib/callbackify.js

```diff
@@ -0,0 +1,17 @@
+'use strict'
+
+module.exports = (fn, opts) => {
+  opts = opts || {}
+  // Min number of non-callback args
+  opts.minArgs = opts.minArgs == null ? 0 : opts.minArgs
+
+  return (...args) => {
+    const cb = args[args.length - 1]
+
+    if (typeof cb !== 'function' || args.length === opts.minArgs) {
+      return fn(...args)
+    }
+
+    fn(...args.slice(0, -1)).then(res => cb(null, res), cb)
+  }
+}
```

Comment:
Isn't there a module that does this?

Reply:
Yeah, not sure exactly why I wrote this anymore, I can probably switch this out for [`callbackify`](https://www.npmjs.com/package/callbackify).

Reply:
Ok, now I remember. I needed something that would allow for variable parameter lengths but would not get confused for these cases:

### 1

```js
await ipfs.pubsub.subscribe(topic, handler)
```
### 2

```js
ipfs.pubsub.subscribe(topic, handler, callback) // 2
```
We want to support: `ipfs.pubsub.subscribe(topic, handler, [options], [callback])`

Our subscribe function is defined like:

```js
function subscribe (topic, handler, options) { /* ... */ }`
```

`callbackify.variadic(subscribe)` gets confused with 1), it thinks `handler` is a callback function.

`callbackify(subscribe)` gets confused in 2) and returns a promise, because it's expecting you to call it with 4 parameters if you want the callback version.

So the version I wrote above allows you to specify the minimum number of non-callback parameters (2 in this case) so that `ipfs.pubsub.subscribe(topic, handler)` returns a promise and `ipfs.pubsub.subscribe(topic, handler, callback)` calls the callback.

I didn't look around for others so there's maybe something out there, maybe I'll send a PR to `callbackify` if I don't find anything. This was just quicker...

Reply:
Check my approach to this here https://github.com/ipfs/js-ipfs/pull/2297/files#diff-18b62220049d9c3c1e6c42df278d0922R75
It's flexible, easy to clean up after we remove callbacks, and promise-nodeify is faster and smarter in some cases (check the readme)